rasa/core/policies/rule_policy.py

```diff
@@ -415,40 +423,66 @@ def _check_prediction(
         domain: Domain,
         interpreter: NaturalLanguageInterpreter,
         gold_action_name: Text,
+        collect_sources: bool,
     ) -> Optional[Text]:
 
         predicted_action_name = self._predict_next_action(tracker, domain, interpreter)
-        if not predicted_action_name or predicted_action_name == gold_action_name:
-            return None
-
         # RulePolicy will always predict active_loop first,
         # but inside loop unhappy path there might be another action
-        if predicted_action_name == tracker.active_loop_name:
+        if (
+            predicted_action_name != gold_action_name
+            and predicted_action_name == tracker.active_loop_name
+        ):
             rasa.core.test.emulate_loop_rejection(tracker)
             predicted_action_name = self._predict_next_action(
                 tracker, domain, interpreter
             )
-            if not predicted_action_name or predicted_action_name == gold_action_name:
-                return None
+
+        if collect_sources:
+            # we need to remember which action should be predicted by the rule
+            # in order to correctly output the names of the contradicting rules
+            rule_name = tracker.sender_id
+            if self._prediction_source in {DEFAULT_RULES, LOOP_RULES}:
+                # the real gold action contradict the one in the rules in this case
+                gold_action_name = predicted_action_name
+                rule_name = self._prediction_source
+
+            self._rules_sources[self._prediction_source].append(
+                (rule_name, gold_action_name)
+            )
+            return
+
+        if not predicted_action_name or predicted_action_name == gold_action_name:
+            return
 
         tracker_type = "rule" if tracker.is_rule_tracker else "story"
-        return (
+        contradicting_rules = {
+            rule_name
+            for rule_name, action_name in self._rules_sources[self._prediction_source]
+            if action_name != gold_action_name
+        }
+
+        error_message = (
             f"- the prediction of the action '{gold_action_name}' in {tracker_type} "
```

Comment:
makes more sense this way, doesn't it?
```suggestion
            f"- the prediction of the action '{predicted_action_name}' in {tracker_type} "
```